### DISCUSSION TITLE
Auto-reload server when saving component file

### DISCUSSION THREAD

**Discussion by twm-jamieb** [django-components/django-components#567] [Q&A] [answered]
Is it just me or does the Django auto-reload watcher not pick up changes in the components .py files to reload the server when making changes? Not the end of the world but pretty annoying when developing.

**Comment by EmilStenstrom** [comment]
@twm-jamieb Could it be the same problem as is being discussed here? https://github.com/EmilStenstrom/django-components/issues/423

**Reply by twm-jamieb** [reply]
I don't think so. That seems to be an issue with the templates caching and I've tried the solution there with no luck. My issue is with the components .py file not triggering StatReloader to restart the server when saving

**Reply by EmilStenstrom** [reply]
Mind trying out the sample project in this repo and see if you can reproduce?

**Reply by twm-jamieb** [reply]
Seems to work in the sample, so must be something I've done in my environment but I can't seem to work out what

**Comment by JuroOravec** [comment]
I had similar issue in my project. Googling around it, I found that Django dev server reloads only when top-level template files change. So if you define each component as a directory, then dev server doesn't reload when the files change.

I used following workaround to make Django reload dev on file changes by adding the files to watch list for our nested templates. See https://stackoverflow.com/a/66673186/9788634.

But, frankly, this still doesn't work 100%, and maybe 5% of the time still I have to hit `CTRL + S` in a non-template file to trigger hot reload.

For context, in this case Django runs in Docker, and dev server is started with `python manage.py runserver`.

```py
from glob import glob
from pathlib import Path

from django.apps import AppConfig
from django.conf import settings
from django.utils.autoreload import autoreload_started


class HomeConfig(AppConfig):
    default_auto_field = 'django.db.models.BigAutoField'
    name = 'home'

    def ready(self):
		# APP_ENV is a custom variable set in settings.py that describes the env
        if settings.APP_ENV == "local":
            watch_list = [
				# Find all nested template files
                *glob('./home/templates/**/*', recursive=True),
                *glob('./project/templates/**/*', recursive=True),
                *glob('./components/**/*', recursive=True),
            ]
            watch_files_for_autoreload(watch_list)


def watch_files_for_autoreload(watch_list: list[str]):
    def autoreload_hook(sender, *args, **kwargs):
        watch = sender.extra_files.add
        for file in watch_list:
            watch(Path(file))

    autoreload_started.connect(autoreload_hook)
```

**Reply by twm-jamieb** [reply]
That works for me, thanks!

**Reply by EmilStenstrom** [reply]
@JuroOravec Maybe this should be added as an optional setup step? Or maybe that we change the defaults so that this works out of the box?

**Reply by JuroOravec** [reply]
Just realized that in that same codebase, I also define all components directories as STATICFILES_DIRS in an attempt to fix it:

```py
# All places where we define Django template
STATICFILES_DIRS = [
    ## BUSINESS LOGIC COMPONENTS:
    # ./<app>/templates/
    *glob_dirs(f"{BASE_DIR}/**/templates"),

    # ./<app>/templates/components/
    # ./<app>/templates/components/<component>/
    # ./<app>/templates/pages/
    # ./<app>/templates/pages/<component>/
    *glob_dirs(f"{BASE_DIR}/**/templates/**/*"),

    ## GENERIC COMPONENTS:
    # ./components/
    *glob_dirs(f"{BASE_DIR}/components"),

    # ./components/<component>/
    *glob_dirs(f"{BASE_DIR}/components/**/*"),
]
```

But I didn't experiment further. @twm-jamieb since you said the first solution works, I assume that's the correct one 😃

@EmilStenstrom  yeah, IMO we could add it as a boolean setting. And the watched files would be all files found recursively in STATICFILES_DIRS.

**Comment by JuroOravec** [comment, marked as answer]
Closing this one, this has been released as part of https://github.com/EmilStenstrom/django-components/releases/tag/0.94. See [Reload dev server on component file changes
](https://github.com/EmilStenstrom/django-components#reload-dev-server-on-component-file-changes)
